### DISCUSSION TITLE
Session Lost for CI 3.1.13 on Page Refresh

### DISCUSSION THREAD

**Discussion by mertdogan** [bcit-ci/CodeIgniter#6247] [General]
Hi. I started a new CI project with CI 3.1.13.

When i refresh page everything on session lost (still on database table) and new session created.

I replaced my new system/library/session folder with CI 3.1.11 (ini_set('session.id', $params['cookie_name']); tricked) and everything works without problem and session not lost.

my simple code at controller is:

```
	public function test(){
		$this->load->library('session');
		var_dump(isset($_SESSION['x'])?$_SESSION['x']:null);
		$_SESSION['x']='123';
		var_dump($_SESSION['x']);
	}
```
when i refresh page for CI3.1.13 than NULL 123 occurs but when i load with CI3.1.11 123 123 occurs on second page load.

I test everything on stackoverflow but can't solve problem.

**Comment by mertdogan** [comment]
my php version is 7.1.19 on IIS and using database driver for session. This is my config setting:
```

$config['sess_driver'] = 'database';
$config['sess_cookie_name'] = 'ODEME_PORTAL_COOKIE';
$config['sess_samesite'] = 'None';
$config['sess_expiration'] = 0;
$config['sess_save_path'] = 'CI_SESSIONS';
$config['sess_match_ip'] = FALSE;
$config['sess_time_to_update'] = 300;
$config['sess_regenerate_destroy'] = FALSE;

/*
|--------------------------------------------------------------------------
| Cookie Related Variables
|--------------------------------------------------------------------------
|
| 'cookie_prefix'   = Set a cookie name prefix if you need to avoid collisions
| 'cookie_domain'   = Set to .your-domain.com for site-wide cookies
| 'cookie_path'     = Typically will be a forward slash
| 'cookie_secure'   = Cookie will only be set if a secure HTTPS connection exists.
| 'cookie_httponly' = Cookie will only be accessible via HTTP(S) (no javascript)
| 'cookie_samesite' = Cookie's samesite attribute (Lax, Strict or None)
|
| Note: These settings (with the exception of 'cookie_prefix' and
|       'cookie_httponly') will also affect sessions.
|
*/
$config['cookie_prefix']	= '';
$config['cookie_domain']	= '';
$config['cookie_path']		= '/';
$config['cookie_secure']	= FALSE;
$config['cookie_httponly'] 	= FALSE;
$config['cookie_samesite'] 	= 'None';

```

**Comment by bunglegrind** [comment]
it sounds like a cookie problem. Please check both the Set-Cookie header in the first response and the Cookie header in your following requests.

**Comment by mertdogan** [comment]
This is request cookie:
![image](https://github.com/bcit-ci/CodeIgniter/assets/1324617/5c17c023-ad4a-4f4b-91f6-4b13ab1438d3)


and this is response cookie:
![image](https://github.com/bcit-ci/CodeIgniter/assets/1324617/416a55e6-ef3e-4cef-80c4-14bbe16f6216)


There is a warning for response cookie as : 
![image](https://github.com/bcit-ci/CodeIgniter/assets/1324617/5fc40218-5372-40f8-bd1c-cd9630f1b1c0)

**Comment by mertdogan** [comment]
After change $config['cookie_secure']	= FALSE; as $config['cookie_secure']	= TRUE; warning gone but session still lost after refresh.

**Comment by mertdogan** [comment]
I think problem is cookie value for session is changing after response. Cookie name for session is ODEME_PORTAL_COOKIE. It's request and response values are different. It must be same for no problem.

**Comment by mertdogan** [comment]
My session values are:
$config['sess_driver'] = 'database';
$config['sess_cookie_name'] = 'ODEME_PORTAL_COOKIE';
$config['sess_samesite'] = 'None';
$config['sess_expiration'] = 0;
$config['sess_save_path'] = 'CI_SESSIONS';
$config['sess_match_ip'] = FALSE;
$config['sess_time_to_update'] = 300;
$config['sess_regenerate_destroy'] = FALSE;

**Comment by bunglegrind** [comment]
maybe the browser is misunderstanding the localhost context? anyway, you should look at the cookies stored in the browser after the first response is arrived (it should match the set-cookie header) and what cookie is sent in the following requests

**Reply by mertdogan** [reply]
I send screenshot of both sent and recieved cookies

**Reply by bunglegrind** [reply]
btw, which version of chrome are you using?

**Reply by mertdogan** [reply]
119.0.6045.124

**Reply by mertdogan** [reply]
It is not about browser. Firefox and edge has same result

**Reply by mertdogan** [reply]
note that: CI3.1.11 session driver (with name to id change trick) session not occures but when i replace system folder with CI3.1.13 problem occurs. I think there is a bug at Codeigniter 3.1.13 for session driver

**Reply by bunglegrind** [reply]
> note that: CI3.1.11 session driver (with name to id change trick) session not occures but when i replace system folder with CI3.1.13 problem occurs. I think there is a bug at Codeigniter 3.1.13 for session driver

no there were a couple of bugs on CI <3.1.12. see the changelog for the details

**Reply by mertdogan** [reply]
I know it bugs but i think there may be another new bug

**Comment by mertdogan** [comment]
There is a new step: when i use files as session driver everything works well. I think there is a session problem about database driver

**Comment by bunglegrind** [comment]
> There is a new step: when i use files as session driver everything works well. I think there is a session problem about database driver


Ok, I tested...It works, even using the database. You have some issue with your configuration. 

config.php:

```php
$config['sess_driver'] = 'database';
$config['sess_cookie_name'] = 'ci_session';
$config['sess_samesite'] = 'None';
$config['sess_expiration'] = 7200;
$config['sess_save_path'] = "ci_sessions";
$config['sess_match_ip'] = FALSE;
$config['sess_time_to_update'] = 300;
$config['sess_regenerate_destroy'] = FALSE;

$config['cookie_prefix']	= '';
$config['cookie_domain']	= '';
$config['cookie_path']		= '/';
$config['cookie_secure']	= FALSE;
$config['cookie_httponly'] 	= FALSE;
$config['cookie_samesite'] 	= 'None';
```

database.php:
```php
$db['default'] = array(
	'dsn'	=> '',
	'hostname' => 'mysql',
	'username' => 'myhumbleuser',
	'password' => 'noneofyourbusiness',
	'database' => 'ci',
	'dbdriver' => 'mysqli',
	'dbprefix' => '',
	'pconnect' => FALSE,
	'db_debug' => FALSE,
	'cache_on' => FALSE,
	'cachedir' => '',
	'char_set' => 'utf8',
	'dbcollat' => 'utf8_general_ci',
	'swap_pre' => '',
	'encrypt' => FALSE,
	'compress' => FALSE,
	'stricton' => FALSE,
	'failover' => array(),
	'save_queries' => TRUE
);
```

table ci_sessions:

```sql
CREATE TABLE IF NOT EXISTS `ci_sessions` (
        `id` varchar(128) NOT NULL,
        `ip_address` varchar(45) NOT NULL,
        `timestamp` int(10) unsigned DEFAULT 0 NOT NULL,
        `data` blob NOT NULL,
        KEY `ci_sessions_timestamp` (`timestamp`)
);
```

**Comment by preciousfocus** [comment]
I have exactly this same issue on moving from C1 3.1.11 to 3.1.13.

Any headway with this?

**Reply by bunglegrind** [reply]
As far as I can see there are no issues to CI. It's something related to your setup/configuration. Please provide more information: config, environment, browser, etc

**Reply by mertdogan** [reply]
Problem still continous but developers don't accept it. I stoped fighting about to fix it and changed my setting as save session to files.

**Reply by bunglegrind** [reply]
have you tested a clean installation? _Divide et impera_

**Reply by mertdogan** [reply]
My installation is a clean installation without any custom code

**Reply by bunglegrind** [reply]
are you sure? I see:
1. html as url suffix  (default is "")
2. A custom name for the cookie
3. I see too many cookies in your first screenshot...there's something more running (try on a private tab)

Clean installation means downloading the zip and editing the configuration the very least in order to have a working instance

...and I suggest you to try Apache instead of IIS, after having setup a working clean installation

**Reply by bunglegrind** [reply]
Moreover, have you checked if the browser is sending the session cookie? It's not clear from the screenshot you provided. Maybe you can use xdebug (or print_r, var_dump, echo, etc...) ...

**Reply by mertdogan** [reply]
I don't mean clean installation as program setting. Your list tems are all about frameworks native settings.

I mean clean installation as there is not any code customization and any custom library for session or database

**Reply by mertdogan** [reply]
I have checked with Firefox and Opera too

**Comment by bunglegrind** [comment]
> I don't mean clean installation as program setting. Your list tems are all about frameworks native settings.

Indeed. The question is: _what are the minimum configuration settings in order to reproduce the issue?_ 

Take in mind I couldn't replicate it. It means...there's something other. 

Regarding the browser: have you tested on a private tab? Does the second request include the Cookie header with the session cookie inside?

Please notice that I've just observed a different behaviour in Chrome with respect to localhost (but no differences between db and files). Use:
```php
$config['sess_samesite'] = 'Lax';//or strict
$config['cookie_secure'] = FALSE;
```

**Reply by mertdogan** [reply]
I tried in private tab too with same result.

I have tried so many settings about sess_samesite and cookie_secure but nothing changed. 

Ok I will build a new project with clear installation (as you meaning) and reply results here.
